Add "E57Simple" API

ptc-jhoerner opened the pull request:
Hi,

this is a port of E57Simple API of the reference library.

Main goal of this is to simplify reading and writing of point clouds with the most common attributes: float x, y, z; float nx, ny, nz; uint8 r, g, b (and any subset of these).

I have removed all the platform specific dependencies from the code, removed the unused code and formatted the code to fit the rest of the library. All duplicate documentation comments have been removed and the doxygen documentation is collected in `E57Simple.h`.

The main changes compared to the original implementation are:
* removed GPS time conversions and the accompanying library (this must be handled on the user side)
* automatically generate GUIDs when writing the data (unless we have user-provided GUIDs)
* make sure all the classes of the simple API are always initialised (gives the user a bit more flexibility when working with metadata classes)
* change data types to fit the most common use-cases -> point coordinates changed to floats (were double), colours are uint8 (were uint16) etc.
* add support for E57_EXT_surface_normals
* use int64_t for all sizes as in the current API
* add equality operators for metadata structs (allows to consistently check whether a field is present or not by comparing to a default initialised struct)
* introduced `Data3DPointsData` to simplify calls to `SetUpData3DPointsData`

You might want to merge this with squashing all the changes. I have kept here the commits to indicate the changes made to the original code.

Cheers,
Jiri

[![image](https://user-images.githubusercontent.com/44769714/83034194-45548080-a038-11ea-9960-de502a30d89d.png)](http://www.ptc.com/)
**Jiri Hörner**
Software Development Engineer, Vuforia

jhoerner@ptc.com

[vuforia.com](http://www.vuforia.com/)

---
Parametric Technology Gesellschaft m.b.H., Operngasse 17-21, 1040 Wien, Austria. Firmensitz: Wien, FN: 111171 m, Firmenbuchgericht: Handelsgericht Wien.
 
The information contained in this email transmission is confidential and may be privileged. If you are not the intended recipient, any use, dissemination, distribution, publication, or copying of the information contained in this email is strictly prohibited.  If you have received this email in error, please immediately notify me by calling the above number and delete the email from your system. Thank you for your co-operation.

Copyright © 2020 PTC Inc. and/or all its affiliates or subsidiaries. All rights reserved

ptc-jhoerner commented on include/E57Simple.h:
```diff
@@ -0,0 +1,888 @@
+      float *normalY{ nullptr }; //!< The Y component of a surface normal vector (E57_EXT_surface_normals).
+      float *normalZ{ nullptr }; //!< The Z component of a surface normal vector (E57_EXT_surface_normals).
+
+      bool ( *pointDataExtension )( ImageFile imf, StructureNode proto, std::vector<SourceDestBuffer> &destBuffers ){
```
I'm not sure whether these extension methods make sense in E57Simple API at all. One can use the current API when writing custom fields or doing anything more complex.

If kept, we should probably consider using std::function.

asmaloney replied:
I would agree with removing them.

asmaloney commented:
As you can tell from the work I put into the original fork, I'm not a fan of "put everything in one file".

Would it be feasible to at least split this into two - one for read (E57SimpleRead) and one for write (E57SimpleWrite)?

asmaloney commented on include/E57Simple.h:
```diff
@@ -0,0 +1,888 @@
+   //! @brief Defines a rigid body transform in cartesian coordinates.
+   struct RigidBodyTransform
+   {
+      e57::Quaternion rotation;     //!< A unit quaternion representing the rotation, R, of the transform
```
Do these (and others throughout this file) need the e57:: namespace?

asmaloney commented on src/E57Simple.cpp:
```diff
@@ -0,0 +1,213 @@
+#include "E57SimpleImpl.h"
+
+using namespace e57;
+using namespace std;
```
I would prefer not to do this. `using namespace std;` is discouraged.

(I understand it is in the older code, but I'd prefer not to add it to new code.)

ptc-jhoerner commented:
Thanks a lot for the feedback and good suggestions. I will look into it next week and get back to you.

asmaloney commented:
Thanks for the submission!

ptc-jhoerner commented:
I have rebased the branch to sync with master.

Aside the changes to address the comments above, I have also added the necessary `E57_DLL` declarations as the shared library changes has been merged now.

The code is now split into 2 files. I have tried to improve the code complexity side as well, but clearly most of the code is still based on the original implementation. I have at least removed some of those extremely long if cases in Reader by employing early exits there. That way it is also more consistent with the Writer.

Many of the other cleanups discussed above went really well. Thanks for those suggestions!

asmaloney commented on include/E57Simple.h:
```diff
@@ -0,0 +1,881 @@
+      //!@{
+
+      //! @brief Returns the file raw E57Root Structure Node
+      StructureNode GetRawE57Root();
```
Should this (and following getters) be const?

asmaloney commented on include/E57Simple.h:
```diff
@@ -0,0 +1,881 @@
+      //!@{
+
+      //! @brief This function returns the file raw E57Root Structure Node
+      StructureNode GetRawE57Root();
```
Should this (and following getters) be const?

asmaloney commented on include/E57Simple.h:
```diff
@@ -0,0 +1,881 @@
+   ////////////////////////////////////////////////////////////////////
+
+   //! @brief Used for reading of the E57 file with E57 Simple API
+   class E57_DLL Reader
```
Would it make sense to put reader in its own header (`E57SimpleReader.h`), writer in its own header (`E57SimpleWriter.h`) and change `E57Simple.h` to `E57SimpleData.h` (or something like that?)

asmaloney commented on src/ReaderImpl.cpp:
```diff
@@ -0,0 +1,1276 @@
+   ReaderImpl::~ReaderImpl()
+   {
+      if ( IsOpen() )
+         Close();
```
Please add braces {} around ifs.

I am going to do this for the whole code base in an upcoming commit. If it's not easy for you to do this, I can use clang-tidy to do it.

ptc-jhoerner replied:
Done with clang-tidy.

asmaloney commented on src/ReaderImpl.h:
```diff
@@ -0,0 +1,111 @@
+   //! most of the functions follows Reader
+   class ReaderImpl
+   {
+   private:
```
Please put `private` after the `public` section (in both new headers).

asmaloney commented on src/ReaderImpl.h:
```diff
@@ -0,0 +1,111 @@
+
+      bool Close();
+
+      bool GetE57Root( E57Root &fileHeader );
```
Should the getters in this class be `const`?

asmaloney commented on src/WriterImpl.cpp:
```diff
@@ -0,0 +1,970 @@
+ */
+
+#include "WriterImpl.h"
+#include "Common.h"
```
Please put in alphabetical order (here and in other files).

ptc-jhoerner replied:
I have moved this one to the group below to keep the header that this is implementing on the top. I have run clang-format with `-sort-includes` over the files, so I hope the order should be fine apart from the grouping.

asmaloney commented on src/WriterImpl.cpp:
```diff
@@ -0,0 +1,970 @@
+      {
+         StructureNode ibox = StructureNode( imf_ );
+
+         //		if( (data3DHeader.indexBounds.rowMinimum != 0) ||
```
Were comments these in the original? Do we need them?

ptc-jhoerner replied:
They came from the original. As far as I can tell it makes sense and it is allowed in the spec to write indexBounds just partially as guarded by those ifs, so I have uncommented them.

asmaloney commented on src/WriterImpl.cpp:
```diff
@@ -0,0 +1,970 @@
+
+      if ( data3DHeader.pointFields.cartesianXField )
+      {
+         if ( data3DHeader.pointFields.pointRangeScaledInteger > 0. )
```
Suggestion: make `pointRangeScale` const above (I think it's const?) and use it to replace `data3DHeader.pointFields.pointRangeScaledInteger` in the following if/elses to make it more readable.

(Same for `data3DHeader.pointFields.pointRangeMinimum` and `data3DHeader.pointFields.pointRangeMaximum`)

ptc-jhoerner replied:
I did this and bit of other improvements in this section as well - reuse of the proto classes etc.

ptc-jhoerner replied:
I hope it is safe to reuse the nodes like this.

ptc-jhoerner replied:
Actually, I have just found a problem with this. Let me fix it first.

ptc-jhoerner replied:
Looking at the code, I don't think it is a good idea to share the same node for multiple fields.

I have reworked the implementation, so now it works in a similar way like the old code, but still avoids most of the repetition.

asmaloney commented on src/WriterImpl.cpp:
```diff
@@ -0,0 +1,970 @@
+         else
+            proto.set( "sphericalAzimuth",
+                       FloatNode( imf_, 0.,
+                                  ( data3DHeader.pointFields.angleScaledInteger < 0. ) ? E57_DOUBLE : E57_SINGLE,
```
Similar to above - use the vars we already created for readability.

asmaloney commented on src/WriterImpl.h:
```diff
@@ -0,0 +1,86 @@
+   //! most of the functions follows Writer
+   class WriterImpl
+   {
+   private:
```
Please put `private` after `public`.

asmaloney commented on src/WriterImpl.h:
```diff
@@ -0,0 +1,86 @@
+      bool WriteData3DGroupsData( int64_t dataIndex, int64_t groupCount, int64_t *idElementValue,
+                                  int64_t *startPointIndex, int64_t *pointCount );
+
+      StructureNode GetRawE57Root();
```
Should these getters be `const`?

asmaloney commented on src/Common.cpp:
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2020 PTC Inc.
```
Suggestion: Forgot to mention that I added info to the README about using [SPDX identifiers](https://spdx.org/ids) to clean up the license texts if you want to use it on your new files (see include/CMakeList.txt files for example).

This whole comment would become:

```cpp
// SPDX-License-Identifier: BSL-1.0
// Copyright 2020 PTC Inc.
```

ptc-jhoerner commented:
Hi, sorry for a slow response from my side. Thanks for the feedback! I'm looking into the changes and I will get back to you tomorrow.

asmaloney commented:
No problem at all Jiri!

I appreciate you taking the time to work on this and that you are willing to incorporate feedback.

ptc-jhoerner commented:
I'm done with the last round of changes. Thanks for all he suggestions. Most of them I have just applied directly.

I took a broader look at the constness of the member methods of both Reader and Writer and I have changed all of them to follow the semantic constness w.r.t. E57 file. Constness is also now the same for the Impl classes.

I have also applied `readability-braces-around-statements` clang-tidy check and I have split the API as suggested, so you might get a bigger diff again.

Cheers,
Jiri
